Reactify Top Nav Menu (kbn_top_nav)

lizozom opened the pull request:
## Summary

Implements https://github.com/elastic/kibana/issues/39981

Updates the top nav of:
 - Dashboard 
 - Visualizations
 - Discover
 - Maps 
 - Timelion
 - Console

## Dev Docs
`Top Nav Menu` is a configurable component that conveniently renders the elements that are often shared by applications: Option menus, query input, filters and time range selection. 

### Show a simple options menu

Define your navigation menu items as an array of `TopNavMenuData` elements and pass it down to the `TopNavMenu` component.

```TSX
import { TopNavMenu, TopNavMenuData } from 'plugins/kibana_react';

const navConfig: TopNavMenuData[] = [{
  id: 'new-item', 
  label: 'New',
  run: () => { 
     showNewItemModal() 
  }
}, {
  id: 'save-item',
  label: 'Save',
  run: () => { 
     if(hasChanges()) {
        saveChanges();
     }
  },
  disableButton: () => {
    return !hasChanges();
  },
  tooltip: () => {
    return !hasChanges() ? 'Nothing to save' : '';
  }
}];

function MyAppComponent(props: Props) {
  return (
    // ... other components rendering ...
    <TopNavMenu
       appName="my-app"
       config={navConfig}
    />
  );
}
```

### Show a full menu with SearchBar

Search Bar is another convenience component that displays `FilterBar`, `QueryBarInput` and `EuiSuperDatePicker` in a standard layout.

This is the full configuration required to render all 3 components.

```TSX
function MyAppComponent(props: Props) {
  const query: Query = {
    query: 'response:200',
    language: 'kuery',
  };
  const filters: Filter[] = [];
  const indexPatterns: IndexPattern[] = [];
  const queryHandler = ({ query, dateRange }) => {};
  const filterHandler = (filters) => {};
  const refreshHandler = ({ isPaused, refreshInterval }) => {};
  return (
    // ... other components rendering ...
    <TopNavMenu
       appName='my-app'
       config={navConfig}
       screenTitle='My App Page'
       store={localStorage}
       indexPatterns={indexPatterns}
       filters={filters}
       query={query}
       onQuerySubmit={queryHandler}
       onFiltersUpdated={filterHandler}
       dateRangeFrom='now-7d'
       dateRangeTo='now'
       isRefreshPaused={false}
       refreshInterval={30}
       onRefreshChange={refreshHandler}
    />
  );
}
```
### Migrating the Angular directive
The `kbn-top-nav` directive is still defined, however, its API has slightly changed. 
While it will be deprecated during the lifetime of 7.x, you may still use it, if you take the following steps.

#### Key attribute is deprecated
Change the `key` attribute of the navigation configuration to `id`. It's a reserved React keyword and hence it's not supported by the `TopNavMenu` component. While the both attributes will work with the directive, it's better to update to `id` right away.
```JS
// Old
const navConfig = [{
  key: 'new-item',
  label: 'new',
  run: () => {},
}];

// New
const navConfig = [{
  id: 'new-item',
  label: 'new',
  run: () => {},
}];
``` 
#### Add a label 
While the `TopNavMenuItem` component still currently falls back to using key\id if `label` is not provided, you should always specify a `label`. This is because `label` is an `i18n` string and `id` is not.

#### Transcluded views are deprecated
Either move them elsewhere or use the SearchBar options.
```JS
// Old
<kbn-top-nav config="topNavConfig">
   <search-bar
      ... search bar options...  
   />
   <my-custom-directive/>
</kbn-top-nav>

// New 
<kbn-top-nav 
    config="topNavConfig"
    ... search bar options...  
/>
<my-custom-directive/>
```
#### Template menu items are deprecated.
Render your views from your app's code and then use the menu's `run` function to toggle visibility.
```JS
// Old
const newTemplateForm = require(....);
const navConfig = [{
  key: 'new-item',
  label: 'new',
  template: newTemplateForm,
}];

// New
const navConfig = [{
  id: 'new-item',
  label: 'new',
  run: () => toggleMyForm,
}];
```

## Screenshots
### Discover
New:
![image](https://user-images.githubusercontent.com/3016806/61452561-b3cea980-a964-11e9-8e55-5c0b74247028.png)

Old:
![image](https://user-images.githubusercontent.com/3016806/60397698-55c65780-9b58-11e9-9e68-696ee4a43a20.png)

 - [x] Menu
 - [x] Filters
 - [x] Query
 - [x] Timepicker 

Still handled in Angular:
 - [x] Hits information

### visualize (configurable components)
New:
![image](https://user-images.githubusercontent.com/3016806/61452721-19bb3100-a965-11e9-9a15-6c92f0640b07.png)

New Markdown (autorefresh only):
![image](https://user-images.githubusercontent.com/3016806/61452593-ce088780-a964-11e9-93ea-5e1fa5884717.png)

New TSVB (timepicker only):
![image](https://user-images.githubusercontent.com/3016806/61452767-37889600-a965-11e9-821f-22c5a7f3eab0.png)

Old:
![image](https://user-images.githubusercontent.com/3016806/60397708-670f6400-9b58-11e9-86b7-4bfaef8194ca.png)

 - [x] Menu
 - [x] Filters
 - [x] Query
 - [x] Timepicker 

Still in Angular:
 - Linked search message (needs design!)

Most visualizations search bar has all 3 features: Timepicker, Filter Bar and Query input (This is true for: Area, Map, Gauge, Table, Goal, Heat map, Horizontal bar, Line, Metric, Pie, Region map, Tag cloud, Vertical bar, Vega). 

However the following have exceptions: (why? @AlonaNadler)
 - **Controls** - has Filter Bar and Timepicker but not Query Input
 - **Markdown** has time autorefresh only.
 - **TSVB** and **Timelion** have Timepicker only.

### dashboard 
New:
![image](https://user-images.githubusercontent.com/3016806/61452819-55ee9180-a965-11e9-8f24-49c8ebab450f.png)

Old:
![image](https://user-images.githubusercontent.com/3016806/60397716-78587080-9b58-11e9-960c-69477439c591.png)
 - [x] Menu
 - [x] Filters
 - [x] Query
 - [x] Timepicker 

### maps (no filter bar)
New:
![image](https://user-images.githubusercontent.com/3016806/61452841-643cad80-a965-11e9-8d80-cb97cbc231d4.png)

Old:
![image](https://user-images.githubusercontent.com/3016806/60397727-93c37b80-9b58-11e9-9135-18c975e85d96.png)
 - [x] Menu
 - [x] Query
 - [x] Timepicker 
 - No filters!

To make it more fun - filters are located inside the layer editing mode, and they use the QueryBar component.

### console (menu only)
New:
![image](https://user-images.githubusercontent.com/3016806/61452871-7880aa80-a965-11e9-8741-ff8cca75dd91.png)

Old:
![image](https://user-images.githubusercontent.com/3016806/60397736-b6ee2b00-9b58-11e9-9a87-96158db13604.png)
 - [x] Menu

### timelion (timepicker only)
New:
![image](https://user-images.githubusercontent.com/3016806/60662210-6df1eb80-9e64-11e9-9de6-f1d2917b8116.png)

Old:
![image](https://user-images.githubusercontent.com/3016806/60397825-b86c2300-9b59-11e9-9359-bdfa1cbbb88a.png)
 - [x] Menu
 - [x] Timepicker
 - No filters \ query!

Still in Angular
 - Query time display  

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5616/)

elasticmachine commented:
Pinging @elastic/kibana-app-arch

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5790/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5798/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5962/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6011/)

lizozom commented on src/legacy/core_plugins/console/public/src/helpers/get_top_nav.ts:
```diff
@@ -36,7 +36,7 @@ export function getTopNavConfig($scope: IScope, toggleHistory: () => {}) {
       }),
       run: () => {
-        toggleHistory();
+        $scope.$evalAsync(toggleHistory);
```
Required when called from React

lizozom commented:
Some additional comments for reviewers:

 - Angular actions called from the new react `TopNavMenu` have to be surrounded by `$scope.$evalAsync(...)` to sync with the Angular digest cycle.
 - QueryBar now supports showing Timepicker without the query input field. I'm not sure the architecture here is optimal, but this may be addressed in an additional PR.
 - The positioning of the visualization linked search message should be addressed by the apps team
 - kbn-top-nav2-helper is a directive helper to help watch the return values if all the disableButton methods for each menu item.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6016/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6019/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6031/)

lizozom commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6287/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6301/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6303/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6316/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6318/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6480/)

lizozom commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6519/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6520/)

lizozom commented:
@timroes I tagged you as a reviewer as there are some changes in this PR in timelion :-D

snide commented on src/legacy/core_plugins/console/public/index.html:
```diff
@@ -1,4 +1,7 @@
-<kbn-top-nav name="console" config="topNavMenu"></kbn-top-nav>
+<kbn-top-nav2 
```
`kbn-top-nav2`? 😸

lizozom replied:
Yeah, the goal is to kill the old directive and then to rename this one back to be kbn-top-nav :)
This is a temporary solution that will not be there when V7.4 is ready.

lizozom replied:
Changed to `kbn-top-nav-v2` :)

snide commented on src/legacy/ui/public/styles/_legacy/_base.scss:
```diff
@@ -59,8 +59,16 @@ input[type="checkbox"],
   }
 
-  > kbn-top-nav {
+  > kbn-top-nav, kbn-top-nav2 {
```
Can this Sass be moved closer to the JS it touches. I don't know that this belongs in base.

All the edits below should then probably be prefixed with `.kbnTopNav` ..etc

lizozom replied:
moved to directive folder \ new component folder

lizozom commented:
@snide thanks for the quick response!
I fixed most of your comments. 
I would love to chat with you directly about the visualizations screen.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6830/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6839/)

cchaos commented:
> **Reload**: We already have this button....

@snide This isn't quite accurate. The "Refresh" button in the globals bar refreshes the current query, while the "Reload" link resets the query to the original saved search. I think we'll still need this button somewhere, but it definitely doesn't need to live in a bar on its own.

As for visualize, I just gave @timroes a mock for how to handle moving the "Linked to Saved Search" to the current place where we establish the index pattern. https://github.com/elastic/kibana/issues/38272#issuecomment-509674634

I'm not sure we want to add more items to the menu bar area as it's possible those menu items might move somewhere else entirely (in the future).

lizozom commented:
@snide @cchaos I wanted to create a navigation menu that will be mostly standard between apps. That's why it has the options menu + search bar \ time picker. It's not built to add custom components into it.

I love the suggestion in the comment https://github.com/elastic/kibana/issues/38272#issuecomment-509674634

snide commented:
> As for visualize, I just gave @timroes a mock for how to handle moving the "Linked to Saved Search" to the current place where we establish the index pattern. #38272 (comment)

This is much better :) Ignore mine!

> The "Refresh" button in the globals bar refreshes the current query, while the "Reload" link resets the query to the original saved search. I think we'll still need this button somewhere, but it definitely doesn't need to live in a bar on its own.

We should likely change the text for it. Agree we can move it (bad idea: maybe next to the title in the breadcrumb). I will say somewhat tangentially that with https://github.com/elastic/kibana/pull/39140 we're nearing a line where we have very similar ways to do these things (saving, reseting and dealing with dirty states) between saved searches / saved queries. Likely there's more going on here we need to think about.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7296/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7437/)

lukeelmers commented on src/legacy/core_plugins/kibana/public/visualize/editor/editor.js:
```diff
@@ -351,10 +351,37 @@ function VisEditor(
+              $scope.showAutoRefreshOnly()) &&
+        !$scope.showQueryInput() &&
+        !$scope.showFilterBar();
+    };
```
Something about the logic in this file is hard for me to grok. I think a lot of it is stemming from the fact that `showAutoRefreshOnly == !showQueryBarTimePIcker`

For example, in `showSearchBarInline`, the `return` statement starts with this conditional:
```js
// if I'm reading the code correctly, this will _always_ evaluate to `true`, so why include it?
return ($scope.showQueryBarTimePicker() || $scope.showAutoRefreshOnly()) && ...etc
```

Or: is it possible for `showQueryBar` to ever return `false`?

Maybe adding more details to your existing comments would help clarify?

lizozom replied:
Holy guac, all visualizations have either time picker or autorefresh only => showQueryBar is always true. Love it!
I simplified the logic significantly and added a comment to the html file.

lukeelmers commented on src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx:
```diff
@@ -547,11 +547,13 @@ export class DashboardAppController {
       dashboardStateManager.setFullScreenMode(false);
-      $scope.kbnTopNav.click(TopNavIds.ADD);
+      // Temp solution for triggering menu click.
+      navActions[TopNavIds.ADD](null, null, null);
```
> // Temp solution for triggering menu click.

How will we know when this can be removed? (Or I guess I'm asking "temporary until when?")

lizozom replied:
Improved comment: 
      /* 
      * Temp solution for triggering menu click.
      * When de-angularizing this code, please call the underlaying action function 
      * directly and not via the top nav object. 
      **/

lukeelmers commented on src/legacy/core_plugins/kibana/public/discover/index.html:
```diff
@@ -1,58 +1,26 @@
+    refresh-interval="refreshInterval.value"
+    on-refresh-change="onRefreshChange"
+  >
+  </kbn-top-nav2>
```
I tested this in Discover, and noticed that the top nav items now stack on narrow screens, instead of the old implementation where they were inline. I'm guessing we want these inline? (Maybe a q for @snide)

### 7.0
![Screenshot 2019-07-15 17 14 18](https://user-images.githubusercontent.com/1608770/61255001-00777200-a724-11e9-9e4b-e63615a86b4a.png)

### This PR
![Screenshot 2019-07-15 17 14 59](https://user-images.githubusercontent.com/1608770/61255019-17b65f80-a724-11e9-8e74-e88bf72c7262.png)

cchaos replied:
There's probably a wrapping EuiFlexGroup that needs `responsive=false`

lizozom replied:
@cchaos @lukeelmers I actually think it's a great behavior, especially if we dare to think about mobile devices.

But I changed it for now.

lukeelmers replied:
FWIW i felt that the other layout was more small-screen-friendly as it took up less vertical space & there are only a few buttons anyway. But I will defer to @cchaos on that :)

lizozom replied:
I'll revert it to the old way it was, so we can proceed with this PR, and if we want, we can address this in a follow up PR

lukeelmers commented on src/legacy/core_plugins/data/public/search/search_bar/components/search_bar.tsx:
```diff
@@ -133,10 +163,10 @@ class SearchBarUI extends Component<Props, State> {
         aria-controls="GlobalFilterGroup"
         aria-expanded={!!this.state.isFiltersVisible}
-        title={`${this.props.filters.length} ${filtersAppliedText} ${clickToShowOrHideText}`}
+        title={`${this.getFilterLength()} ${filtersAppliedText} ${clickToShowOrHideText}`}
```
What happens to this string if `getFilterLength` returns `void`? (Is that even possible?)

lizozom replied:
Fixed

lukeelmers replied:
I guess what I meant here was that `getFilterLength` looks like this:
```ts
  private getFilterLength() {
    if (this.props.showFilterBar && this.props.filters) {
      return this.props.filters.length;
    }
  }
```

so if the conditional evaluates to false, the function won't return (`void`). what happens then in your template string?

```js
`${this.getFilterLength()} ${filtersAppliedText} ${clickToShowOrHideText}`
```

is `getFilterLength` just going to be an empty string? this is mostly a question, i honestly don't know and haven't tested

lizozom replied:
The message is used like this:

![image](https://user-images.githubusercontent.com/3016806/61443128-92b08d80-a951-11e9-80ab-2443764fefc2.png)

So you're right. I fixed to show only "select to show\hide" when filter count is 0.

lukeelmers commented on src/legacy/core_plugins/data/public/search/search_bar/components/search_bar.test.tsx:
```diff
@@ -0,0 +1,206 @@
+  };
+});
+
+const noop = () => {
```
nit: could alternatively use `jest.fn()` in the tests instead of creating a `noop` fn here

lukeelmers commented on src/legacy/core_plugins/data/public/search/search_bar/components/search_bar.test.tsx:
```diff
@@ -0,0 +1,206 @@
+    expect(component.find(QUERY_BAR).length).toBe(0);
+  });
+
+  it('Should render filter bar, when required fields are probided', () => {
```
```suggestion
  it('Should render filter bar, when required fields are provided', () => {
```

lukeelmers commented on src/legacy/core_plugins/data/public/query/query_bar/components/query_bar.test.tsx:
```diff
@@ -102,4 +104,104 @@ describe('QueryBar', () => {
+    expect(component.find(TIMEPICKER_SELECTOR).length).toBe(0);
+  });
+
+  it('Should render timepicker with options', () => {
```
This doesn't seem to be testing anything different than the first test?

lizozom replied:
It passes in the timepicker options explicitly.

lukeelmers commented on src/legacy/ui/public/kbn_top_nav/kbn_top_nav2.js:
```diff
@@ -0,0 +1,114 @@
+
+const module = uiModules.get('kibana');
+
+module.directive('kbnTopNav2', () => {
```
We should add this to the angular migration guide too ❤️

lizozom replied:
Done

lukeelmers commented on src/legacy/core_plugins/kibana_react/public/index.ts:
```diff
@@ -0,0 +1,26 @@
+// TODO these are imports from the old plugin world.
+// Once the new platform is ready, they can get removed
+// and handled by the platform itself in the setup method
+// of the ExpressionExectorService
```
Naive question: Would it still work to put this stuff in `src/plugins/kibana_react` rather than creating a duplicate legacy plugin?

The only two places I see it importing from the legacy world are `SearchBar` in the `data` plugin, and some scss files. So can you still statically import from those places in the new platform? Or are linting rules preventing this?

lizozom replied:
a. We don't have a way to load css files from the new platform plugin yet.
b. In a following PR I also moved SearchBar into `kibana_react`

Anyway, I prefer to move it to new platform in a separate PR, after I merge the follow up PRs.
Hope it's cool :-P

lukeelmers replied:
Ah yeah forgot about the css thing. 👍

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7907/)

lizozom commented on src/legacy/core_plugins/console/public/src/helpers/get_top_nav.ts:
```diff
@@ -36,7 +36,7 @@ export function getTopNavConfig($scope: IScope, toggleHistory: () => {}) {
       }),
       run: () => {
-        toggleHistory();
+        $scope.$evalAsync(toggleHistory);
```
Done

lizozom commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/top_nav_menu_item.tsx:
```diff
@@ -0,0 +1,57 @@
+      typeof menuData.disableButton === 'function'
+        ? menuData.disableButton()
+        : menuData.disableButton;
+    return val || false;
```
Done in a bit of a different way.
Instead of passing in a `data` object, I made `TopNavMenuItem` Props extend `TopNavMenuData`

```
interface Props extends TopNavMenuData {
  onClick: (key: string, action: TopNavMenuAction, target?: any) => void;
}
```

And then I could just use the react defaultProps:
```
TopNavMenuItem.defaultProps = {
  disableButton: false,
}
```

lizozom commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/top_nav_menu.tsx:
```diff
@@ -0,0 +1,123 @@
+  showSearchBarInline?: boolean;
+};
+
+export function TopNavMenu(props: Props) {
```
done

lizozom commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/top_nav_menu.test.tsx:
```diff
@@ -0,0 +1,88 @@
+    },
+  ];
+
+  beforeEach(() => {});
```
You have to admit its one of my best line of code.
So simple yet so fundamentally true!

Deleted tho :-D

lizozom commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/top_nav_menu_item.test.tsx:
```diff
@@ -0,0 +1,104 @@
+    component.simulate('click', event);
+    expect(clickHandler).toHaveBeenCalledTimes(0);
+  });
+});
```
done

lizozom commented on x-pack/legacy/plugins/maps/public/index.js:
```diff
@@ -22,7 +22,6 @@ import { capabilities } from 'ui/capabilities';
 import chrome from 'ui/chrome';
 import routes from 'ui/routes';
 import 'ui/kbn_top_nav';
-import 'ui/angular-bootstrap'; // required for kbn-top-nav button tooltips
```
I added tooltip support to top nav now!

lizozom commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/_index.scss:
```diff
@@ -0,0 +1,7 @@
+.topNavMenu__wrapper {
+  z-index: 5;
```
Yes, the second one is the old top nav, which still exists in 2 places.
It will disappear along with the folder in ui/public.

lizozom commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/top_nav_menu_data.tsx:
```diff
@@ -0,0 +1,31 @@
+ * under the License.
+ */
+
+export type TopNavMenuAction = (menuItem: any, navController: any, anchorElement: any) => void;
```
I was able to simplify this significantly thanks to your comment and remove those unneeded options.
Thanks!

lizozom commented on test/functional/services/inspector.js:
```diff
@@ -30,7 +30,7 @@ export function InspectorProvider({ getService }) {
   return new class Inspector {
     async getIsEnabled() {
-      const ariaDisabled = await testSubjects.getAttribute('openInspectorButton', 'aria-disabled');
+      const ariaDisabled = await testSubjects.getAttribute('openInspectorButton', 'disabled');
```
This specific line checks if the _top nav inspector button_ is disabled.
That button is now an EuiEmptyButton, and EUI chose to only set the **disabled** attribute (which makes sense according [to this](https://stackoverflow.com/a/38059790))

I see no harm in doing it this way, do you?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7952/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7963/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7982/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8057/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8134/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8160/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8217/)

cchaos commented on src/legacy/core_plugins/kibana/public/discover/index.html:
```diff
@@ -122,6 +90,15 @@ <h1 id="kui_local_breadcrumb" class="kuiLocalBreadcrumb" ng-if="opts.savedSearch
             </button>
 
+            <div class="dscResultCount">
+              <span data-test-subj="discoverQueryHits" class="kuiLocalBreadcrumb__emphasis">{{(hits || 0) | number:0}}</span>
```
Instead of using this legacy class, just make the element a `strong` tag
```suggestion
              <strong data-test-subj="discoverQueryHits">{{(hits || 0) | number:0}}</strong>
```

cchaos commented on src/legacy/core_plugins/kibana/public/discover/index.html:
```diff
@@ -122,6 +90,15 @@ <h1 id="kui_local_breadcrumb" class="kuiLocalBreadcrumb" ng-if="opts.savedSearch
+                i18n-default-message="{hits, plural, one {hit} other {hits}}"
+                i18n-values="{ hits }"
+              ></span>
+            </div>
```
Can you add the refresh/reset button to this container and center the contents like so:

<img width="620" alt="Screen Shot 2019-07-19 at 15 13 19 PM" src="https://user-images.githubusercontent.com/549577/61559535-e72e3880-aa37-11e9-8203-2855ccb1d066.png">

lizozom replied:
The original button was relevant only to saved searches.
I made the button show only for saved searches in the location you suggested.

Did you mean to also move the main reset button (the blue one in the search bar)? If so, I think it should be a part of a wider redesign :)

cchaos commented on src/legacy/core_plugins/kibana/public/visualize/editor/editor.html:
```diff
@@ -1,55 +1,60 @@
-        data-transclude-slot="bottomRow"
-        ng-show="chrome.getVisible()"
-        class="fullWidth"
+  <!-- Linked search. -->
```
For the linked search, I had given @timroes [a mock](https://github.com/elastic/kibana/issues/38272#issuecomment-509674634) to move this to where the current index-pattern is shown like so:

![image](https://user-images.githubusercontent.com/549577/61559750-69b6f800-aa38-11e9-9b23-9cd68fdd5734.png)

lizozom replied:
Great to have this linked.

cchaos commented on src/legacy/core_plugins/kibana_react/public/top_nav_menu/top_nav_menu.tsx:
```diff
@@ -0,0 +1,132 @@
+  function renderLayout() {
+    if (props.showSearchBarInline) {
+      return (
+        <div className="topNavMenu__wrapper">
```
Can you prefix all these classes with `kbn` example:`kbnTopNavMenu__wrapper` that way we know what scope this lies in (Kibana global)?

lizozom replied:
done

cchaos commented on src/legacy/ui/public/styles/_legacy/_base.scss:
```diff
@@ -59,8 +59,8 @@ input[type="checkbox"],
 
-  > kbn-top-nav {
-    z-index: 5;
+  .globalQueryBar {
```
This should go in this file: https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/data/public/filter/filter_bar/_global_filter_group.scss

Not legacy styles.

lizozom replied:
done

cchaos commented on x-pack/legacy/plugins/maps/public/angular/map_controller.js:
```diff
@@ -332,7 +332,7 @@ app.controller('GisMapController', ($scope, $route, kbnUrl, localStorage, AppSta
   $scope.showDatePicker = true; // used by query-bar directive to enable timepikcer in query bar
   $scope.topNavMenu = [{
-    key: i18n.translate('xpack.maps.mapController.fullScreenButtonLabel', {
+    id: i18n.translate('xpack.maps.mapController.fullScreenButtonLabel', {
```
More translations on id's

lizozom replied:
The reason for this is that historically ID can be used as a label.
I aligned it to have an id and a translated label. Thanks for the catch!

lizozom commented on src/legacy/core_plugins/kibana/public/dashboard/top_nav/get_top_nav_config.ts:
```diff
@@ -144,7 +144,7 @@ function getCloneConfig(action: NavAction) {
 function getAddConfig(action: NavAction) {
   return {
-    key: i18n.translate('kbn.dashboard.topNave.addButtonAriaLabel', {
+    id: i18n.translate('kbn.dashboard.topNave.addButtonAriaLabel', {
```
fixed

lizozom replied:
fixed

lizozom commented on src/legacy/core_plugins/data/public/search/search_bar/components/search_bar.tsx:
```diff
@@ -114,85 +175,62 @@ class SearchBarUI extends Component<Props, State> {
+          appName={this.props.appName}
+          indexPatterns={this.props.indexPatterns}
+          store={this.props.store!}
+          prepend={this.props.showFilterBar ? this.getFilterTriggerButton() : ''}
```
fixed

lizozom commented on src/legacy/core_plugins/data/public/search/search_bar/components/search_bar.test.tsx:
```diff
@@ -0,0 +1,204 @@
+        showFilterBar={false}
+        filters={[]}
+        onFiltersUpdated={noop}
+        showDatePicker={false}
```
Date picker will show if not disabled explicitly, because it can work without any options.
I know it's not optimal, and I intend to address this in a separate issue that will re-think the API of search bar \ filter bar \ query bar compoents.

lizozom commented:
@cchaos Thanks!
I addressed all of your comments, including removing the showSearchBarInline option, which I wasn't excited about in the first place anyway :-).
Looking forward to seeing how we work the new design out.

cchaos commented on src/legacy/core_plugins/kibana/public/discover/_discover.scss:
```diff
@@ -36,8 +36,18 @@ discover-app {
 
 .dscResultCount {
+  display: flex;
+  justify-content: center;
```
This is fine, but as an FYI, it's easier to just set it as `text-align: center` since it's just text inside. It's more likely to be fine across browsers.

cchaos commented on src/legacy/core_plugins/kibana/public/discover/_discover.scss:
```diff
@@ -36,8 +36,18 @@ discover-app {
+    padding-left: $euiSizeXS; 
+  }
+
+  > button {
```
Try not to use plain element selectors
```suggestion
  > .kuiLink {
```

cchaos commented on src/legacy/core_plugins/kibana/public/discover/index.html:
```diff
@@ -91,12 +91,25 @@
+                type="button"	
+                id="reload_saved_search"	
+                aria-label="{{::'kbn.discover.reloadSavedSearchAriaLabel' | i18n: {defaultMessage: 'Reload saved search'} }}"	
+                tooltip="{{::'kbn.discover.reloadSavedSearchTooltip' | i18n: {defaultMessage: 'Reload saved search'} }}"	
```
Unnecessary to add `tooltip` and `aria-label` because the button's text is fully visible.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8537/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8594/)

cchaos commented:
@lizozom I just happened to notice one thing in Maps. It looks like in full screen the top nav bar doesn't go away anymore:

<img src="https://d.pr/free/i/DunmxU+" width="50%" />

lizozom commented:
@cchaos Added fix in https://github.com/elastic/kibana/pull/41672

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10215/)
